### PR TITLE
KAFKA-14568: Move FetchDataInfo and related to storage module

### PR DESCRIPTION
**Pull request by fvaleri**
Part of KAFKA-14470: Move log layer to storage module.

### CODE REVIEW

**Comment by fvaleri** [comment]
@ijuma @satishd fyi

File: storage/src/main/java/org/apache/kafka/server/log/internals/FetchIsolation.java [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.server.log.internals;
+
+import org.apache.kafka.common.IsolationLevel;
+import org.apache.kafka.common.requests.FetchRequest;
+import org.apache.kafka.common.utils.FetchRequestUtils;
+
+public interface FetchIsolation {
```
Comment (ijuma) [inline storage/src/main/java/org/apache/kafka/server/log/internals/FetchIsolation.java]: This should be an enum.

Comment (fvaleri) [reply storage/src/main/java/org/apache/kafka/server/log/internals/FetchIsolation.java]: I thought about that, but then I used `javap` to decompile the class and this is how it comes out.

Comment (ijuma) [reply storage/src/main/java/org/apache/kafka/server/log/internals/FetchIsolation.java]: A Java enum is the right way to represent `FetchIsolation`. In Scala, it's often emulated with a sealed trait and case objects. I don't understand the `javap` relevance here.

Comment (fvaleri) [reply storage/src/main/java/org/apache/kafka/server/log/internals/FetchIsolation.java]: The relevance is that an enum is compiled into a class, while the sealed trait into an interface. Anyway, I'll try to change into an enum.

Comment (ijuma) [reply storage/src/main/java/org/apache/kafka/server/log/internals/FetchIsolation.java]: And it's better if it's compiled to a class generally speaking (lower number of potential performance pitfalls).

Comment (fvaleri) [reply storage/src/main/java/org/apache/kafka/server/log/internals/FetchIsolation.java]: Ok, thanks. Should be fine now.

File: core/src/main/scala/kafka/log/UnifiedLog.scala [hunk 2]
```diff
@@ -1249,12 +1249,10 @@ class UnifiedLog(@volatile var logStartOffset: Long,
...
     checkLogStartOffset(startOffset)
-    val maxOffsetMetadata = isolation match {
-      case FetchLogEnd => localLog.logEndOffsetMetadata
-      case FetchHighWatermark => fetchHighWatermarkMetadata
-      case FetchTxnCommitted => fetchLastStableOffsetMetadata
-    }
-    localLog.read(startOffset, maxLength, minOneMessage, maxOffsetMetadata, isolation == FetchTxnCommitted)
+    var maxOffsetMetadata = localLog.logEndOffsetMetadata
+    if (isolation == FetchIsolation.FETCH_HIGH_WATERMARK) maxOffsetMetadata = fetchHighWatermarkMetadata
+    else if (isolation == FetchIsolation.FETCH_TXN_COMMITTED) maxOffsetMetadata = fetchLastStableOffsetMetadata
```
Comment (ijuma) [inline core/src/main/scala/kafka/log/UnifiedLog.scala]: I think you can keep the `match` here as it was.

File: storage/src/main/java/org/apache/kafka/server/log/internals/FetchIsolation.java [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+package org.apache.kafka.server.log.internals;
+
+import org.apache.kafka.common.IsolationLevel;
+import org.apache.kafka.common.requests.FetchRequest;
+import org.apache.kafka.common.utils.FetchRequestUtils;
+
+public enum FetchIsolation {
+    FETCH_LOG_END,
+    FETCH_HIGH_WATERMARK,
+    FETCH_TXN_COMMITTED;
```
Comment (ijuma) [inline storage/src/main/java/org/apache/kafka/server/log/internals/FetchIsolation.java]: We can probably remove the `FETCH` prefix.

**Comment by fvaleri** [comment]
Let's wait for CI builds, but it should be good now. Thanks for the review.

File: clients/src/main/java/org/apache/kafka/common/utils/FetchRequestUtils.java [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.common.utils;
+
+public class FetchRequestUtils {
```
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/common/utils/FetchRequestUtils.java]: I would just add these methods to the existing `FetchRequest` class.

File: core/src/main/scala/kafka/cluster/Partition.scala [hunk 5]
```diff
@@ -1189,7 +1189,7 @@ class Partition(val topicPartition: TopicPartition,
    * @param minOneMessage whether to ensure that at least one complete message is returned
    * @param updateFetchState true if the Fetch should update replica state (only applies to follower fetches)
    * @return [[LogReadInfo]] containing the fetched records or the diverging epoch if present
-   * @throws NotLeaderOrFollowerException if this node is not the current leader and [[FetchParams.fetchOnlyLeader]]
+   * @throws NotLeaderOrFollowerException if this node is not the current leader and `FetchParams.fetchOnlyLeader`
```
Comment (ijuma) [inline core/src/main/scala/kafka/cluster/Partition.scala:1192]: Are these changes related to this PR?

File: core/src/main/scala/kafka/log/LocalLog.scala [hunk 6]
```diff
@@ -1005,12 +1005,12 @@ object LocalLog extends Logging {
 
   private[log] def emptyFetchDataInfo(fetchOffsetMetadata: LogOffsetMetadata,
                                       includeAbortedTxns: Boolean): FetchDataInfo = {
-    val abortedTransactions =
-      if (includeAbortedTxns) Some(List.empty[FetchResponseData.AbortedTransaction])
-      else None
-    FetchDataInfo(fetchOffsetMetadata,
+    val abortedTransactions: Optional[java.util.List[FetchResponseData.AbortedTransaction]] =
+      if (includeAbortedTxns) Optional.of(List.empty[FetchResponseData.AbortedTransaction].asJava)
```
Comment (ijuma) [inline core/src/main/scala/kafka/log/LocalLog.scala]: We can pass `Collections.emptyList` instead of creating a Scala list and then converting to Java.

File: core/src/main/scala/kafka/server/ReplicaManager.scala [hunk 7]
```diff
@@ -123,12 +123,12 @@ case class LogReadResult(info: FetchDataInfo,
     this.info.records,
     this.divergingEpoch,
     this.lastStableOffset,
-    this.info.abortedTransactions,
+    if (this.info.abortedTransactions.isPresent) Some(this.info.abortedTransactions.get().asScala.toList) else None,
```
Comment (ijuma) [inline core/src/main/scala/kafka/server/ReplicaManager.scala]: The last `toList` results in a copy of the collection, we'd want to avoid that.

File: storage/src/main/java/org/apache/kafka/server/log/internals/FetchIsolation.java [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+import org.apache.kafka.common.IsolationLevel;
+import org.apache.kafka.common.requests.FetchRequest;
+import org.apache.kafka.common.utils.FetchRequestUtils;
+
+public enum FetchIsolation {
+    LOG_END,
+    HIGH_WATERMARK,
+    TXN_COMMITTED;
+
+    public static FetchIsolation apply(FetchRequest request) {
```
Comment (ijuma) [inline storage/src/main/java/org/apache/kafka/server/log/internals/FetchIsolation.java]: I'd call this method and the other methods `of`. That's more common for Java.

File: storage/src/main/java/org/apache/kafka/server/log/internals/FetchParams.java [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+import java.util.Objects;
+import java.util.Optional;
+
+public class FetchParams {
+    private final short requestVersion;
+    private final int replicaId;
+    private final long maxWaitMs;
+    private final int minBytes;
+    private final int maxBytes;
+    private final FetchIsolation isolation;
```
Comment (ijuma) [inline storage/src/main/java/org/apache/kafka/server/log/internals/FetchParams.java]: We can make these fields public since they're immutable. The accessors don't add much value in such cases.

File: storage/src/main/java/org/apache/kafka/server/log/internals/FetchParams.java [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+    private Optional<ClientMetadata> clientMetadata;
```
Comment (ijuma) [inline storage/src/main/java/org/apache/kafka/server/log/internals/FetchParams.java]: How come this isn't final?

File: storage/src/main/java/org/apache/kafka/server/log/internals/FetchParams.java [hunk 11]
```diff
@@ -0,0 +1,125 @@
...
+    public FetchIsolation isolation() {
+        return isolation;
+    }
+
+    public Optional<ClientMetadata> clientMetadata() {
+        return clientMetadata;
+    }
+
+    @Override
+    public boolean equals(Object o) {
```
Comment (ijuma) [inline storage/src/main/java/org/apache/kafka/server/log/internals/FetchParams.java:69]: Do we actually need to implement `equals` and `hashCode` for this? Is it used as the key of a map or something like that? I suspect it may not be needed and it was only there because it comes for free with case classes. Same question for other classes where we have implemented these methods.

File: storage/src/main/java/org/apache/kafka/server/log/internals/FetchDataInfo.java [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+    public FetchDataInfo(LogOffsetMetadata fetchOffsetMetadata,
+                         Records records,
+                         Optional<List<FetchResponseData.AbortedTransaction>> abortedTransactions) {
+        this(fetchOffsetMetadata, records, false, abortedTransactions);
+    }
+
+    public FetchDataInfo(LogOffsetMetadata fetchOffsetMetadata,
+                         Records records,
+                         boolean firstEntryIncomplete,
+                         Optional<List<FetchResponseData.AbortedTransaction>> abortedTransactions) {
```
Comment (ijuma) [inline storage/src/main/java/org/apache/kafka/server/log/internals/FetchDataInfo.java:40]: We have a lot of constructors, do we need these all?

File: core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala [hunk 13]
```diff
@@ -2312,26 +2313,26 @@ class ReplicaManagerTest {
...
+                               replicaId: Int,
+                               fetchInfos: Seq[(TopicIdPartition, PartitionData)],
+                               responseCallback: Seq[(TopicIdPartition, FetchPartitionData)] => Unit,
+                               requestVersion: Short = ApiKeys.FETCH.latestVersion,
+                               maxWaitMs: Long = 0,
+                               minBytes: Int = 1,
+                               maxBytes: Int = 1024 * 1024,
+                               quota: ReplicaQuota = UnboundedQuota,
+                               isolation: FetchIsolation = FetchIsolation.LOG_END,
+                               clientMetadata: Option[ClientMetadata] = None
```
Comment (ijuma) [inline core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala]: Why this indent change?

**Review by ijuma — COMMENTED** [review]
I took a pass and left some comments. Note that the fetch path is performance sensitive, so we have to be careful with collection conversions and other such things.

File: storage/src/main/java/org/apache/kafka/server/log/internals/FetchDataInfo.java (hunk 12 above)
Comment (fvaleri) [reply storage/src/main/java/org/apache/kafka/server/log/internals/FetchDataInfo.java:40]: I added them based on existing code usage.

File: core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala (hunk 13 above)
Comment (fvaleri) [reply core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala]: Reverting.

File: storage/src/main/java/org/apache/kafka/server/log/internals/FetchParams.java (hunk 10 above)
Comment (fvaleri) [reply storage/src/main/java/org/apache/kafka/server/log/internals/FetchParams.java]: Fixing.

File: storage/src/main/java/org/apache/kafka/server/log/internals/FetchParams.java (hunk 11 above)
Comment (fvaleri) [reply storage/src/main/java/org/apache/kafka/server/log/internals/FetchParams.java:69]: Yes, they are needed for the following tests:

- ReplicaAlterLogDirsThreadTest.shouldReplaceCurrentLogDirWhenCaughtUp
- ReplicaAlterLogDirsThreadTest.shouldUpdateLeaderEpochAfterFencedEpochError

They both call `mockFetchFromCurrentLog` that uses the equal matcher on a instance of `FetchParams`.

File: core/src/main/scala/kafka/cluster/Partition.scala (hunk 5 above)
Comment (fvaleri) [reply core/src/main/scala/kafka/cluster/Partition.scala:1192]: TBH, I haven't found a way to link a Java method from scaladoc. Do you have any suggestion here?

File: core/src/main/scala/kafka/server/ReplicaManager.scala (hunk 7 above)
Comment (fvaleri) [reply core/src/main/scala/kafka/server/ReplicaManager.scala]: Agree. I also moved `FetchPartitionData` to get rid of this.

**Comment by fvaleri** [comment]
> Note that the fetch path is performance sensitive, so we have to be careful with collection conversions and other such things.

Yes and thanks for the detailed reviews. Much appreciated.

**Review by ijuma — APPROVED** [review]
I pushed a few clean-ups and it LGTM now. @fvaleri Are you good with them too?

File: core/src/test/scala/unit/kafka/server/ReplicaAlterLogDirsThreadTest.scala [hunk 14]
```diff
@@ -276,25 +276,22 @@ class ReplicaAlterLogDirsThreadTest {
...
 
-    println(expectedFetchParams)
-
     when(replicaManager.fetchMessages(
       params = ArgumentMatchers.eq(expectedFetchParams),
       fetchInfos = ArgumentMatchers.eq(Seq(topicIdPartition -> requestData)),
       quota = ArgumentMatchers.eq(UnboundedQuota),
       responseCallback = callbackCaptor.capture(),
     )).thenAnswer(_ => {
-      println("Did we get the callback?")
```
Comment (ijuma) [inline core/src/test/scala/unit/kafka/server/ReplicaAlterLogDirsThreadTest.scala:297]: Deleted two stray `println` here.

File: storage/src/main/java/org/apache/kafka/server/log/internals/FetchDataInfo.java (hunk 12 above)
Comment (ijuma) [reply storage/src/main/java/org/apache/kafka/server/log/internals/FetchDataInfo.java:40]: I reduced them a bit to make it more manageable.

File: storage/src/main/java/org/apache/kafka/server/log/internals/FetchParams.java (hunk 11 above)
Comment (ijuma) [reply storage/src/main/java/org/apache/kafka/server/log/internals/FetchParams.java:69]: Fair enough, I adjusted the implementations a bit.

**Comment by ijuma** [comment]
JDK 8 and 11 builds passed, 17 had 2 unrelated failures.

**Comment by fvaleri** [comment]
> I pushed a few clean-ups and it LGTM now. @fvaleri Are you good with them too?

Yes, good. Thanks.
